{% extends "base.html" %}

{% macro form_entry(thing, tooltip) %}
  <li>
    <span id="formlabel"
          tooltip="{{ tooltip }}">
    {{ thing.label }}
    </span>
    {{ thing }}
    {% if thing.errors %}
    <div class=errors>
      {% for error in thing.errors %}
      {{ error }}<br>
      {% endfor %}
    </div>
    {% endif %}
{% endmacro %}

{% macro infotitle(text) %}
  <span id="infotitle">{{ text }}</span>
{% endmacro %}

{% block content %}
<h1><span>phylogenize</span></h1>
<h2>v0.2 alpha</h2>

{% with messages = get_flashed_messages() %}
  {% if messages %}
    <div class="errors">
      {% for message in messages %}
      {{ message }} <br>
      {% endfor %}
    </div>
  {% endif %}
{% endwith %}

		<div id="infobox"><p>


      {{ infotitle("Purpose") }}
      This web tool allows you to associate microbial prevalence (or
      specificity for a given environment) with individual microbial genes,
      starting from microbial relative abundances and sample labels. The method
      is described in
      <a href="https://www.biorxiv.org/content/early/2018/03/22/189795">
        Bradley, Nayfach, and Pollard (2018)</a>.<p>

      {{ infotitle("Input") }}
      phylogenize requires two user-provided tables. The first is a table of
      reads or relative abundances, with microbes in rows and samples in
      columns. The second is a table of sample annotations, where sample IDs
      are rows and the columns give environment and dataset labels.
      Environments usually give the location of the sample
      (e.g., human skin), but can also refer to healthy vs. disease in a
      case-control study.

      You can provide these tables either separately, as tab-delimited files,
      or as a single BIOM-formatted file.
      
      If you are calculating prevalence, you can also
      provide multiple dataset labels to combine information over multiple
      experiments. These experiments will then be weighted equally when
      calculating prevalence. An example relative abundance file can be found
      <a href="bodysite-averaged.tab">here</a>, and an example metadata file
      can be found <a href="bodysite-metadata.tab">here</a>.<p>

      {{ infotitle("Species abundance table options") }}
      The OTU/species table can either be generated by applying MIDAS to
      metagenomic data or by applying DADA2 to 16S data. If uploading 16S data,
      you should first use the DADA2 package to make species level assignments
      based on the PATRIC database. To do so, simply download the PATRIC <a
        href="ftp://ftp.patricbrc.org/16s_RNA/16s_RNA_PATRIC.frn.gz">16S FASTA
        file</a> and use it with the <tt>assignSpecies()</tt> function
      (see the DADA2 <a href="https://benjjneb.github.io/dada2/tutorial.html">
      tutorial</a>, under "Assign taxonomy"; we use <tt>assignSpecies()</tt>
      instead of <tt>assignTaxonomy()</tt> to take advantage of exact matching).
      Write the output to a tab-delimited file and upload it as the "DADA2
      mapping file" below. If using MIDAS, row names should be MIDAS IDs. Be
      sure to select the appropriate database.<p>
      
      <span id="infotitle">Phenotype</span>
      You can find associations either with prevalence, how commonly a
      particular microbe is observed in a given environment, or with
      environmental specificity, a measure of how predictive a given microbe is
      for an environment. The environment you are interested in should be
      entered under the "which environment" heading, exactly as it appears in
      the metadata file. (If you like, you can also skip this step and provide
      your own phenotype, such as drug resistance.) Power users can also
      control how likely each environment should be a priori by selecting an
      "environmental prior type" or uploading their own priors.<p>

			<span id="infotitle">Other options</span>
      You can find more information about any field by mousing over the title.<p>

  </div>

		<form name="UserUpload"
      action=""
      method="POST"
      autocomplete="off"
      enctype="multipart/form-data"
      novalidate>

      {{ form.hidden_tag() }}

      <ul>
        {{ form_entry(form.abundances, "Tab-separated values. Rows should be
        MIDAS species IDs or PATRIC genome IDs.") }} {{
        form_entry(form.metadata, "Tab-separated values. Rows should be
        samples, and columns should be as in example file above.") }} {{
        form_entry(form.biomfile, "BIOM file (combined abundance table and
        sample annotation table). Make sure your sample annotation table has
        &quot;env&quot; and &quot;dataset&quot; annotations.") }} {{
        form_entry(form.datatype, "Select whether data were calculated from
        applying MIDAS to metagenomic data, or DADA2 to 16S data.") }} {{
        form_entry(form.database, "Select the version of the database that you
        used to get relative abundances.") }} {{ form_entry(form.phenotype,
        "Select whether you want to associate gene families with prevalence or
        with environmental specificity scores.") }} {{
        form_entry(form.which_envir, "Enter the name of the environment in
        which you want to calculate prevalence or specificity.") }}
      </ul>
      {{ form.recaptcha }}<p>
      <input type="submit" value="Submit" formmethod="post"><br>

    </form>

<h3>Patrick H. Bradley, Pollard lab</h3>

{% endblock %}
