{% extends "base.html" %}

{% macro form_entry(thing, tooltip, upload) %}
		<label class="control-label">
    <span class="formlabel"
          tooltip="{{ tooltip }}">
    {{ thing.label }}
    </span>
		</label>
		{% if not upload %}
    	{{ thing(class_="form-control") }}
		{% else %}
			<label class="form-control">
			<span class="btn btn-outline-primary">Browse</span>
    	{{ thing(class_="hidden") }}
			</label>
		<p>
		{% endif %}
    {% if thing.errors %}
    <div class=errors>
      {% for error in thing.errors %}
      {{ error }}<br>
      {% endfor %}
    </div>
    {% endif %}
{% endmacro %}

{% macro infotitle(text) %}
  <span class="infotitle">{{ text }}</span>
{% endmacro %}

{% block content %}
{% with messages = get_flashed_messages() %}
  {% if messages %}
    <div class="errors">
      {% for message in messages %}
      {{ message }} <br>
      {% endfor %}
    </div>
  {% endif %}
{% endwith %}

<div class="jumbotron jumbotron-fluid jumbotron-fancy">
  <div class="container">
    <h1 class="display-4">Tutorial</h1>
    <p class="lead">Learn how to upload and analyze your own data here.</p>
  </div>
</div>


<section><p/>
    <div class="container">
        <div class="card border-secondary w-50">
            <div class="card-body">
                <h5 class="card-title">Table of contents</h5>
                <ul>
                    <li>
                        <a href="#what-is-phylogenize">
                            What is <em>phylogenize</em>?
                        </a>
                    </li>
                    <li>
                        <a href="#using-phylogenize">
                            Using <em>phylogenize</em>
                        </a>
                        <ul>
                            <li>
                                <a href="#overview">
                                    Overview
                                </a>
                            </li>
                            <li>
                                <a href="#data">
                                    Data
                                    <ul>
                                        <li>
                                            <a href="#shotgun">
                                                Shotgun
                                            </a>
                                        </li>
                                        <li>
                                            <a href="#sixteens">
                                                16S
                                            </a>
                                        </li>
                                    </ul>
                                </a>
                            </li>
                            <li>
                                <a href="#metadata">
                                    Metadata
                                </a>
                            </li>
                            <li>
                                <a href="#file-formats">
                                    File formats
                                    <ul>
                                        <li>
                                            <a href="#tabular">
                                                Tabular
                                            </a>
                                        </li>
                                        <li>
                                            <a href="#biom">
                                                BIOM
                                            </a>
                                        </li>
                                    </ul>
                                </a>
                            </li>
                            <li>
                                <a href="#choosing-a-phenotype">
                                    Choosing a phenotype
                                </a>
                            </li>
                            <li>
                                <a href="#choosing-an-environment">
                                    Choosing an environment
                                </a>
                            </li>
                            <li>
                                <a href="#additional-options">
                                    Additional options
                                </a>
                            </li>
                            <li>
                                <a href="#interpreting-results">
                                    Interpreting results
                                </a>
                            </li>
                        </ul>
                    </li>
                    <li>
                        <a href="#examples">
                            Examples
                        </a>
                    </li>
                    <li>
                        <a href="#troubleshooting">
                            Troubleshooting
                        </a>
                    </li>
                    <li>
                        <a href="#running-phylogenize-locally">
                            Running <em>phylogenize</em> locally
                        </a>
                    </li>
                    <li>
                        <a href="#citing-phylogenize">
                            Citing <em>phylogenize</em>
                        </a>
                    </li>
                </ul>
            </div>
        </div>
        <p/>

        <h1 id="what-is-phylogenize">What is <em>phylogenize</em>?</h1>

        <strong><p><em>phylogenize</em> is a web tool that allows users to link
            microbial genes to environments, accounting for phylogeny.</strong></p>

        <p>More specifically, given community composition data,
        <em>phylogenize</em> links genes in microbial genomes to either microbial
        prevalence in, or specificity for, a given environment, while also
        taking into account an important potential confounder: the phylogenetic
        relationships between microbes. The method is described fully in <a
        href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006242">Bradley,
        Nayfach, and Pollard (2018)</a>.</p>

        <h1 id="using-phylogenize">Using <em>phylogenize</em></h1>

        <h2 id="overview">Overview</h2>

        <p>To use <em>phylogenize</em>, you need to upload the following
        things:</p>
        <ul>
            <li>
                <strong>Community composition data</strong>, either from 16S or
                shotgun sequencing data.
            </li>
            <li>
                <strong>Sample metadata</strong>, which maps sample IDs in your
                community composition data to environments.
            </li>
        </ul>

        <p>You also need to select the following:</p>
        <ul>
            <li>
                <strong>Whether to link genes to prevalence or
                specificity</strong>, which are defined as follows:
                <ul>
                    <li><i>prevalence</i>: how frequently is a microbe observed in a
                        given environment?</li>
                    <li><i>specificity</i>: how specific is a microbe for a
                        given environment compared to all others?</li>
                    <li>Note: to calculate specificity, there must be more than
                        one environment present.</li>
                </ul>
            </li>
            <li>
                <strong>A choice of environment</strong>: prevalence and
                specificity are defined with respect to a given environment, like
                &quot;soil&quot; or &quot;stool&quot; or &quot;marine.&quot;
            </li>
        </ul>

        <p>Here is a visual explainer of the <em>phylogenize</em> user interface:</p>

        <center>
        <div class="w-75 figure">
            <img class="inline" src="{{ url_for('static',
                        filename="phylogenize-form-example.png") }}"
                        alt="Example of form" />
            <p class="caption">Example of form</p>
        </div>
        </center>

        <h2 id="#data">Data</h2>

        <p><em>phylogenize</em> requires taxonomic abundances. These can either
            be derived from shotgun data or 16S data, and can be <strong>read
            counts, relative abundances, or presence-absence data</strong>.</p>

        <p>More specifically, your data should be provided as a matrix where:
            <ul>
                <li>each row is a species ID (shotgun data) or a sequence variant (16S data)</li>
                <li>each column is a sample ID of your choosing</li>
                <li>each entry is any of the following:</li>
                <ul>
                    <li>a number of reads</li>
                    <li>a relative abundance</li>
                    <li>0 for absence and 1 for presence</li>
                </ul>
            </li>
            </ul>
        </p>

        <h3 id="shotgun">Shotgun data</h3>

        <p>Shotgun data should be processed with
            <a href="https://github.com/snayfach/MIDAS">MIDAS</a>, with database
            version either v1.2 (recommended) or v1.0 (deprecated). You do not need
            to run the full MIDAS pipeline; you will just need to do metagenomic
            species profiling as described
            <a href="https://github.com/snayfach/MIDAS/blob/master/docs/species.md">here</a>.
            The output file you need should be called <code>species_profile.txt
            </code>.</p>

        <h3 id="sixteens">16S data</h3>

        <p>16S data should be processed using a <strong>denoising
            algorithm</strong>. Two common options are <a
            href="https://github.com/biocore/deblur">Deblur</a> and <a
            href="https://benjjneb.github.io/dada2/">DADA2</a>. Both of these
            algorithms will result in abundances for &quot;amplicon sequence
            variants&quot; (ASVs), which correspond to the distinct DNA
            sequences that are obtained by denoising.</p>

        <p><em>phylogenize</em> works by mapping these ASVs back to <a
            href="https://www.github.com/snayfach/MIDAS">MIDAS</a> genome
            clusters using the fast aligner <a
            href="https://github.com/knights-lab/BURST">BURST</a>. The default
            threshold for matching a genome cluster is 98.5% sequence identity,
            since we found that this gives a reasonable sensitivity-specificity
            threshold. Ambiguous matches are discarded, and reads/abundances
            mapping to the same genome cluster are summed together. (In
            practice, though, we have never actually seen ambiguous
            matches.)</p>

        <h2 id="metadata">Metadata</h2>

        <p>Your metadata should be provided in the form of a matrix where:
            <ul>
                <li>each row corresponds to a sample in your data;</li>
                <li>there are at least the following two columns:</li>
                <ul>
                    <li>a column containing the <strong>sample IDs</strong>,
                        matching the sample IDs in your data (named
                        <code>sample</code> for tabular data and
                        <code>#SampleID</code> for BIOM data);</li>
                    <li>a column contains the <strong>environment
                        labels</strong> corresponding to each sample (by default
                        called <code>env</code>, but you can specify a different
                        column name)</li>
                </ul>
                <li>and, optionally:
                    <ul>
                        <li>a column containing the <strong>batch or
                            dataset</strong> to which each sample belongs. This
                            is used for adjusting prevalence estimates across
                            studies with unequal numbers of samples (by default
                            called <code>dataset</code>, but you can specify a
                            different column name).</li>
                        <li>If you only have one dataset, you don't need to
                            specify a dataset column. Just select 1 under
                            &quot;Number of datasets&quot;.</li>
                        <li><em>Note</em>: right now, you can only calculate
                            specificity if there is just one dataset.</li>
                    </ul>
                </li>
            </ul>
        </p>

        <h2 id="file-formats">File formats</h2>

        <p>Your data and metadata should be in either <strong>tabular</strong>
            or <strong>BIOM</strong> format. Tabular format is simpler, but
            takes up more space. The maximum file upload is 35M, so if your data
            are larger, you will need to convert them to BIOM format (see below)
            or, alternatively, to run <em>phylogenize</em> locally.</p>

        <p>You can upload either:
            <ul>
                <li>
                    two tab-delimited files (one data file and one metadata
                    file);
                </li>
                <li>
                    one BIOM file (containing data and metadata);
                </li>
                <li>
                    or one tab-delimited metadata file and one BIOM-formatted
                    abundance file.
                </li>
            </ul>
        </p>

        <h3 id="tabular">Tabular format data</h3>

        <p>Tabular data and metadata files should look like this:</p>

        <div class="figure">
            <img class="inline"
                 src="{{ url_for('static', filename="data-corner.png") }}"
                 alt="Top corner of data" />
            <p class="caption">Top corner of data</p>
        </div>

        <div class="figure">
            <img class="inline"
                 src="{{ url_for('static', filename="metadata-corner.png") }}"
                 alt="Top corner of metadata" />
            <p class="caption">Top corner of metadata</p>
        </div>

        <p>Note that the metadata table needs to have a column labeled <em>sample</em>
            (matching the columns of your data table).</p>
        <p>The table should also
            giving the environment of each sample (by default,
            <em>phylogenize</em> looks for a column named &quot;env&quot;, but
            you can change this in the &quot;Environment column&quot; field).</p>
        <p>If your data represents samples from different datasets, batches, or
            studies, one column should also say which dataset, batch, or study
            each sample is taken from (by default: &quot;dataset&quot;).</p>

        <h3 id="biom">BIOM format data</h3>

        <p><a href="http://biom-format.org/">BIOM format</a> data is more
            compact for tabular data and is more suitable for larger datasets.
            This is because 1. BIOM represents matrices that are mostly zeroes
            (i.e., &quot;sparse&quot;) more efficiently, and most taxonomic
            matrices are sparse, and 2. BIOM files can use a binary
            representation (HDF5) that takes up less space than plain text.</p>

        <p>A BIOM file can contain multiple tables, so you will only need to
            upload one file (though if you prefer, you can upload a separate
            metadata file as described above). This one file needs to have the
            following tables:</p>

        <ul>
            <li><strong>Observation matrix</strong> or &quot;OTU
                        table&quot;. This is where the community composition
                        data will go. <i>Despite the name, rows should actually
                        be either MIDAS species IDs or amplicon sequence
                        variants (ASVs), not OTUs.</i>
            </li>
            <li>
                <strong>Sample metadata matrix</strong> (unless uploading a
                separate tabular file).
            </li>
        </ul>

        <p>The sample metadata matrix should include a column with environment
            annotations (by default called <em>env</em>) and, if there are
            multiple datasets, dataset annotations (by default called
            <em>dataset</em>). Whatever these columns are named, they should
            match the names you provide in the &quot;Environment column&quot;
            and &quot;Dataset column&quot; fields respectively, just as with
            tabular data.</p>
        </ul>

        <p>To convert your tabular data into a single BIOM file,
            <a href="http://biom-format.org/index.html">install the
                <code>biom</code> utility</a>. Then run the conversion as
            follows (adapted from the biom-format.org documentation):</p>

        <pre><code>biom convert -i your_data.tab -o your_data_and_metadata.biom --to-hdf5 --table-type=&quot;OTU table&quot; --sample-metadata-fp metadata.tab</code></pre>

        <p>Replace <code>your_data.tab</code> and <code>metadata.tab</code> with
            the names of your tab-delimited data files and, optionally,
            <code>your_data_and_metadata</code> with the name of your dataset.
            <i>Important note</i>: before doing this step, edit your metadata
            file and rename the &quot;sample&quot; column to
            &quot;#SampleID&quot;, otherwise the metadata won't actually be
            added (!) and <i>phylogenize</i> will throw a &quot;metadata not
            found&quot; error.</p>

        <p>For more help using the <code>biom</code> utility, try looking at the
            pages entitled
            <a href="http://biom-format.org/documentation/biom_conversion.html">
                &quot;Converting between file formats&quot;</a> and
            <a href="http://biom-format.org/documentation/adding_metadata.html">
                &quot;Adding sample and observation metadata to biom
                files&quot;</a>.</p>

        <h2 id="choosing-a-phenotype">Choosing a phenotype</h2>

        <p><em>phylogenize</em> can associate genes with either microbial
            <em>prevalence</em> or microbial <em>specificity</em> for a given
            environment. Prevalence gives how often a microbe is found in an
            environment. Specificity compares this prevalence to prevalences in
            all other provided environments. The following toy example gives an
            intuition for what this means:</p>
        <div class="figure">
            <img src="{{ url_for('static', filename="healthy-sick.png") }}"
                 class="inline"
                 alt="Cartoon of prevalence and specificity calculations" />
            <p class="caption">
                Cartoon of prevalence and specificity calculations
            </p>
        </div>
        <p>Microbe 1 is detected (black boxes) often in both samples from
            healthy (A-G) and sick (H-N) individuals. Microbe 2 is detected more
            frequently in healthy than sick samples, and microbe 3 is detected
            relatively seldom in either. This means that microbes 1 and 2 will
            have <em>high prevalence</em> in the &quot;healthy&quot;
            environment, while microbe 3 will have <em>low prevalence</em>.
            However, microbe 2 is the only microbe to have <em>high
            specificity</em> for &quot;healthy&quot; over &quot;sick,&quot;
            because of the difference in prevalence between these
            environments.</p>

        <p>Prevalence will tend to capture both cosmopolitan microbes and ones
            that are particularly well-adapted to a given environment.
            Specificity is useful for contrasting two sites (like gut vs. skin)
            or states (healthy vs. sick).</p>

        <p>The exact details of how prevalence and specificity are calculated
            can be read in our
            <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006242">
                paper</a>.
            Basically, we perform additional steps to re-weight datasets equally
            (for prevalence), to make sure that small numbers of observations
            don't skew the results (for specificity), and to transform these
            quantities into something approximately normally-distributed
            (both).</p>

        <p><b>Note</b>: it is possible with low numbers of samples that no microbes
            will be specific to a given environment, in which case
            <em>phylogenize</em> will return an error because there is not enough
            signal to compute specificity.</p>

        <p><b>Note</b>: while you do not have to rarefy your data (i.e., resample to
            the same read depth), you should make sure that read depth does not
            systematically differ across environments. If it does, then
            rarefying may be a good idea because some microbes will simply
            appear more by chance in deeper-sequenced data, and if read depth is
            correlated with environment, you may get spurious associations.</p>

        <p><b>Note:</b> in order to be considered, each environment and/or
            dataset must have at least two samples associated with it. This is
            because it doesn't really make sense to calculate prevalence based on a
            single presence/absence value (and because specificity, in turn,
            depends partly on calculating prevalence in each environment).</p>

        <h2 id="choosing-an-environment">Choosing an environment</h2>

        <p>Simply enter the name of the environment for which you are interested
            in calculating a phenotype. This should match an entry in the
            <em>env</em> column of the sample metadata (or the other column you
            named in the &quot;Environment column&quot; field) exactly. At least
            two samples per environment are necessary.</p>

        <h2 id="additional-options">Additional options</h2>

        <p>&quot;Data type&quot; is relatively self-explanatory: is your data
            16S amplicon sequencing data, or shotgun metagenomics data processed
            with MIDAS?</p>

        <p>For most people, the MIDAS database version will be the latest, v1.2.
            We also provide the option to use a previous version of the
            database, since species are named differently in MIDAS v1.2 and
            v1.0, but for most people the default (1.2) will be appropriate.</p>

        <h2 id="interpreting-results">Interpreting results</h2>

        <p>You can see what a sample results page looks like
            <a href="{{ url_for('static', filename='phylogenize-report-demo.html') }}">here</a>.</p>

        <p>Your results will be delivered at a URL that should look like:
            <code>www.phylogenize.org/results/&lt;result_id&gt;</code> with a
            long string replacing <code>&lt;result_id&gt;</code>. This is where
            you should eventually be able to pick up your results. We do not
            make the result IDs searchable, so it is a good idea to bookmark or
            otherwise note this URL.</p>

        <p>Anyone with this URL will be able to see your results, so only share
            it with members of your team and collaborators. (Also, even though we
            don't make these URLs publicly searchable, you still should not upload
            any data with personally-identifiable information, sensitive personal
            information, or any other information that needs to remain 100%
            completely secure.)</p>

        <p>The results have two components. The first is an HTML report
            summarizing and visualizing the data. The second is a .tgz file
            (compressed archive) that should contain the full set of results,
            including the following files:</p>
        <ul>
            <li><code>phenotype.tab</code>: Tab-delimited file giving the
                calculated phenotype values (i.e., prevalence or specificity)
                for every microbe.</li>
            <li><code>pos-sig-thresholded.csv</code>: Comma-delimited file
                giving all FIGfam gene families significantly positively
                associated with the phenotype, within a given phylum. This
                file also contains descriptions of the significant
                genes.</li>
            <li><code>all-results.csv</code>: Comma-delimited file giving
                the effect size and uncorrected p-value for the association
                of every FIGfam gene family in every phylum with the
                calculated phenotype. This can be useful if, for example,
                you want to apply your own threshold for significance or a
                different p-value correction method, or if you want to
                further filter your results by effect size.</li>
            <li><code>enr-table.csv</code>: FDR-corrected SEED subsystem
                enrichments for the significantly positively associated
                FIGfam gene families. These are pre-computed for three
                gene-wise levels of significance: &quot;strong&quot; (0.05),
                &quot;med&quot; (0.1), and &quot;weak&quot; (0.25). SEED
                subsystems significant at an FDR of 25% are returned.</li>
            <li><code>enr-overlaps.csv</code> and
                <code>enr-overlaps-sorted.csv</code>: These files give the
                individual significant FIGfams underlying any SEED subsystem
                enrichments, which can aid interpretation.</li>
            <li><code>progress.txt</code> and <code>stderr.txt</code>: These
                are the outputs of <em>phylogenize</em>; if your job
                finished without an error message, you probably don't need
                to refer to them.</li>
            </li>
        </ul>

        <p>To make <em>phylogenize</em> available for other users, results will
            be deleted approximately 7 days after completion, so it is a good
            idea to save the report and results to disk before then.</p>

        <h1 id="examples">Examples</h1>

        <h2 id="hmp">Human Microbiome Project</h2>

        <p>If you would like to take <i>phylogenize</i> for a test drive, we
            have provided a <a href="{{ url_for('static',
            filename="hmp-16s.biom") }}">BIOM dataset</a> containing 16S data
            and metadata from the Human Microbiome Project, re-analyzed with
            DADA2. (The data are read counts, merged by individual. The various
            regions sequenced of the 16S gene are provided in a single file.
            Reads corresponding to different samples from the same individual
            have been summed.) <b>Note</b>: this is a large dataset, so it may
            take a long time to run (on the order of an hour). We have also 
						provided a <a href="{{ url_for('static', filename="hmp-16s-dada2-bacteroidetes.tab") }}">reduced dataset</a> in tabular form with its own <a href="{{ url_for('static', filename="hmp-16s-metadata-bacteroidetes.tab") }}"> metadata </a> consisting of only the Bacteroidetes, 
				    which should run faster, particularly when calculating specificity.</p>

        <p>You will have to provide an environment to calculate either
            specificity or prevalence. The various environments represented in
            this dataset are:
            <ul><li>Anterior nares</li>
                <li>Attached/Keratinized gingiva</li>
                <li>Buccal mucosa</li>
                <li>Hard palate</li>
                <li>L_Antecubital fossa</li>
                <li>Mid vagina</li>
                <li>Palatine Tonsils</li>
                <li>Posterior fornix</li>
                <li>R_Antecubital fossa</li>
                <li>Retroauricular crease</li>
                <li>Saliva</li>
                <li>Stool</li>
                <li>Subgingival plaque</li>
                <li>Supragingival plaque</li>
                <li>Throat</li>
                <li>Tongue dorsum</li>
                <li>Vaginal introitus</li>
            </ul>
        </p>

        <p>A sample report generated on these data, containing the results for
            associating genes with specificity for the "Stool" environment, can
            be found <a href="{{ url_for('static',
            filename='hmp-stool-phylo/index.html') }}">here</a>.</p>

        <h2 id="hmp">Earth Microbiome Project</h2>

        <p>In our <a
            href="https://www.biorxiv.org/content/10.1101/425231v2">preprint</a>,
            we use data from the Earth Microbiome Project together with
            <em>phylogenize</em> to identify microbial genes associated with the
            plant rhizosphere. The full EMP data are very large, so we only
            provide the report file
            <a href="{{ url_for('static', filename='plant-rhizosphere-phylo/index.html') }}">here</a>.</p>

        <h1 id="troubleshooting">Troubleshooting</h1>

        <p>Here are a few areas where you may run into trouble:</p>

        <ul>
            <li><em>phylogenize</em> appears to be stuck a little over halfway through.
                <ul>
                    <li>This is probably totally normal; the progress bar is
                        approximate and this is likely to be the step where
                        associations are actually calculated. If you check
                        back, you should eventually see more subtle signs of
                        progress under &quot;monitor warnings/errors.
                        &quot;</li>
                </ul>
            </li>
            <li><em>phylogenize</em> doesn't appear to be running my
                job: the bar is stuck at 0%.
                <ul>
                    <li>Because <em>phylogenize</em> takes a lot of memory
                        to run, only one job can run at a time. If you
                        reload the page and look under &quot;monitor
                        warnings/errors&quot; you should see information
                        about how many other jobs are in the queue. If you
                        don't see this information and the job is still
                        stuck at 0% for a prolonged period of time, contact
                        support (see About/Contact in the navigation
                        bar).</li>
                </ul>
            </li>
            <li><em>phylogenize</em> gave me an error that I don't
                understand.
                <ul>
                    <li>Check that your data tables satisfy the above
                        requirements (in particular, are your metadata
                        columns named correctly, and do your sample IDs
                        match up?).</li>
                    <li>If your job ran out of memory, try converting it
                        to BIOM format (see above).</li>

                </ul>
            </li>
        </ul>

        <p>If there's still a problem, e-mail the webmaster
            (see About/Contact in the navigation bar). Add
            your result ID (the string of characters after
            &quot;results&quot; in the URL) and attach your
            input files and the output that appears when you
            click &quot;monitor warnings/errors&quot;.</p>

        <p>We also encourage you to report a bug in <em>phylogenize</em> using
            the <a href="https://bitbucket.org/pbradz/phylogenize/issues">issue
            tracker</a>.</p>

        <h1 id="running-phylogenize-locally">Running <em>phylogenize</em> locally</h1>

        <p>In addition to a web tool, <em>phylogenize</em> can also be run on a
            laptop, desktop, or server. In particular, if you are getting an
            &quot;out of memory&quot; error, you have a lot of jobs, or you have
            a more specific use case that is not covered by the server, we
            recommend that you download <em>phylogenize</em> from our code
            repository at <a
            href="https://www.bitbucket.com/pbradz/phylogenize">
            https://bitbucket.com/pbradz/phylogenize</a>. </p>

        <p>The core of <em>phylogenize</em> is an R package, with a web app written
            in Python using Flask. The web app provides a user interface and a
            scheduling system (using Beanstalk). <em>phylogenize</em> can also
            be run in <a href="https://qiime2.org/">QIIME2</a> using the plugin
            <a
                href="https://bitbucket.org/pbradz/q2-phylogenize">q2-phylogenize</a>.
        </p>
        <p>
            This means that there are three ways to run <em>phylogenize</em>
            locally:
            <ul>
                <li><strong>In R</strong>, calling functions from the R package directly;</li>
                <li><strong>In QIIME2</strong>, using the plugin <a href="https://bitbucket.org/pbradz/q2-phylogenize">q2-phylogenize</a>;</li>
                <li><strong>In a web browser</strong>, by running the web application locally.</li>
            </ul>
            For most people, the first two approaches are probably the easiest.
            The third option is recommended only if you want multiple people to
            be able to run <em>phylogenize</em> on the same local machine.
        </p>
        <p>
            Instructions on how to use <em>phylogenize</em> as an R package or
            through a local web server are found on its <a
            href="https://bitbucket.org/pbradz/phylogenize">Bitbucket site</a>.
            Instructions on how to use <em>phylogenize</em> with QIIME2 can be
            found on the <a href="https://bitbucket.org/pbradz/q2-phylogenize">
            q2-phylogenize</a> Bitbucket repository.

        <h1 id="citing-phylogenize">Citing <em>phylogenize</em></h1>

        <p>
            Bradley PH and Pollard KS, "phylogenize: a web tool to identify
            microbial genes underlying environment associations." In review.<p>
            Bradley PH, Nayfach S, and Pollard KS, "Phylogeny-corrected
            identification of microbial gene families relevant to human gut
            colonization." PLOS Comput Biol 14 (8), e1006242.<p>
    </div>
</section>



{% endblock %}
