{% extends "base.html" %}

{% macro form_entry(thing, tooltip, upload) %}
		<label class="control-label">
    <span class="formlabel"
          tooltip="{{ tooltip }}">
    {{ thing.label }}
    </span>
		</label>
		{% if not upload %}
    	{{ thing(class_="form-control") }}
		{% else %}
			<label class="form-control">
			<span class="btn btn-outline-primary">Browse</span>
    	{{ thing(class_="hidden") }}
			</label>
		<p>
		{% endif %}
    {% if thing.errors %}
    <div class=errors>
      {% for error in thing.errors %}
      {{ error }}<br>
      {% endfor %}
    </div>
    {% endif %}
{% endmacro %}

{% macro infotitle(text) %}
  <span class="infotitle">{{ text }}</span>
{% endmacro %}

{% block content %}
{% with messages = get_flashed_messages() %}
  {% if messages %}
    <div class="errors">
      {% for message in messages %}
      {{ message }} <br>
      {% endfor %}
    </div>
  {% endif %}
{% endwith %}

<div class="jumbotron jumbotron-fluid jumbotron-fancy">
  <div class="container">
    <h1 class="display-4">Tutorial</h1>
    <p class="lead">Learn how to upload and analyze your own data here.</p>
  </div>
</div>

<section><p>
<div class="container">

					{{ infotitle("Purpose") }}
					This web tool allows you to associate microbial prevalence (or
					specificity for a given environment) with individual microbial genes,
					starting from microbial relative abundances and sample labels. The method is described in Bradley, Nayfach, and Pollard (2018).<p>
					<p>

					{{ infotitle("Input") }}
					phylogenize requires two user-provided tables. The first is a table of
					reads or relative abundances, with microbes in rows and samples in
					columns. The second is a table of sample annotations, where sample IDs
					are rows and the columns give environment and dataset labels.
					Environments usually give the location of the sample
					(e.g., human skin), but can also refer to healthy vs. disease in a
					case-control study.<p>

					You can provide these tables either separately, as tab-delimited files,
					or as a single BIOM-formatted file.<p>

					If you are calculating prevalence, you can also
					provide multiple dataset labels to combine information over multiple
					experiments. These experiments will then be weighted equally when
					calculating prevalence. An example relative abundance file can be found
					<a href="bodysite-averaged.tab">here</a>, and an example metadata file
					can be found <a href="bodysite-metadata.tab">here</a>.<p>

					{{ infotitle("Species abundance table options") }}
					The OTU/species table can either be generated by applying MIDAS to
					metagenomic data or by applying DADA2 to 16S data. If uploading 16S data,
					you should first use the DADA2 package to make species level assignments
					based on the PATRIC database. To do so, simply download the PATRIC <a
										 href="ftp://ftp.patricbrc.org/16s_RNA/16s_RNA_PATRIC.frn.gz">16S FASTA
										 file</a> and use it with the <tt>assignSpecies()</tt> function
									 (see the DADA2 <a href="https://benjjneb.github.io/dada2/tutorial.html">
										 tutorial</a>, under "Assign taxonomy"; we use <tt>assignSpecies()</tt>
									 instead of <tt>assignTaxonomy()</tt> to take advantage of exact matching).
									 Write the output to a tab-delimited file and upload it as the "DADA2
									 mapping file" below. If using MIDAS, row names should be MIDAS IDs. Be
									 sure to select the appropriate database.<p>

									 {{ infotitle("Phenotype") }}	
									 You can find associations either with prevalence, how commonly a
									 particular microbe is observed in a given environment, or with
									 environmental specificity, a measure of how predictive a given microbe is
									 for an environment. The environment you are interested in should be
									 entered under the "which environment" heading, exactly as it appears in
									 the metadata file. (If you like, you can also skip this step and provide
									 your own phenotype, such as drug resistance.) Power users can also
									 control how likely each environment should be a priori by selecting an
									 "environmental prior type" or uploading their own priors.<p>

									 {{ infotitle("Other options") }}
									 You can find more information about any field by mousing over the title.
		</div>
</section>



{% endblock %}
