<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8">
		<link rel="stylesheet" href="phylo.css">  
		<link href="https://fonts.googleapis.com/css?family=Barlow" rel="stylesheet">
    <title>phylogenize</title>
  </head>
  <body>
    <h1 ><span
          >phylogenize</span></h1>
		<h2>v0.1 alpha</h2>

		<div id="infobox"><p>
			<span id="infotitle">Purpose</span> This web tool allows you to associate microbial prevalence (or specificity for a given environment) with individual microbial genes, starting from microbial relative abundances and sample labels. The method is described in <a href="https://www.biorxiv.org/content/early/2018/03/22/189795">Bradley, Nayfach, and Pollard (2018)</a>.<p>
			<span id="infotitle">Input</span> This tool requires two tab-delimited tables. The first is an OTU or species table of relative abundances, with microbes in rows and samples in columns. The second is a table of metadata, where samples are rows and the columns give environment and dataset labels. These can be labeled any way you choose. Environments usually give the location of the sample (e.g., human skin), but can also refer to healthy vs. disease in a case-control study. If you are calculating prevalence, you can also provide multiple dataset labels to combine information over multiple experiments. These experiments will then be weighted equally when calculating prevalence. An example relative abundance file can be found <a href="bodysite-averaged.tab">here</a>, and an example metadata file can be found <a href="bodysite-metadata.tab">here</a>.<p>
			<span id="infotitle">OTU/species table options</span> The OTU/species table can either be generated by applying MIDAS to metagenomic data or by applying DADA2 to 16S data. If uploading 16S data, you should first use the DADA2 package to make species level assignments based on the PATRIC database. To do so, simply download the PATRIC <a href="ftp://ftp.patricbrc.org/16s_RNA/16s_RNA_PATRIC.frn.gz">16S FASTA file</a> and use it with the <tt>assignSpecies()</tt> function (see the DADA2 <a href="https://benjjneb.github.io/dada2/tutorial.html">tutorial</a>, under "Assign taxonomy"; we use <tt>assignSpecies()</tt> instead of <tt>assignTaxonomy()</tt> to take advantage of exact matching). Write the output to a tab-delimited file and upload it as the "DADA2 mapping file" below. If using MIDAS, row names should be MIDAS IDs. Be sure to select the appropriate database.<p>
			<span id="infotitle">Phenotype</span> You can find associations either with prevalence, how commonly a particular microbe is observed in a given environment, or with environmental specificity, a measure of how predictive a given microbe is for an environment. The environment you are interested in should be entered under the "which environment" heading, exactly as it appears in the metadata file. (If you like, you can also skip this step and provide your own phenotype, such as drug resistance.) Power users can also control how likely each environment should be a priori by selecting an "environmental prior type" or uploading their own priors.<p>
			<span id="infotitle">Other options</span> You can find more information about any field by mousing over the title.<p>
		</div>

		<form name="UserUpload" action="uploader.py" method="POST" autocomplete="off" enctype="multipart/form-data">
			<p>
			<span id="bigformlabel">Input data</span>
      <ul >
				<li><span id="formlabel" tooltip="Tab-separated values. Rows should be MIDAS species IDs or PATRIC genome IDs.">OTU/species table</span>
            <input name="abundance_file" required="required" accept="text/tab-separated-values"
              type="file"></li>
				<li><span id="formlabel" tooltip="Tab-separated values. Rows should be samples, and columns should be as in example file above.">Metadata table</span> <input
              name="metadata_file" required="required" accept="text/tab-delimited-values"
              type="file"></li>
				<li><span id="formlabel" tooltip="Select whether data were calculated from applying MIDAS to metagenomic data, or DADA2 to 16S data.">Data type</span> <input
              name="dtype" required="required" value="midas" type="radio" checked="checked">MIDAS
            <input name="dtype" required="required" value="16S" 
              type="radio">16S</li>
				<li><span id="formlabel" tooltip="Tab-separated values. Should be the output of DADA2 assignSpecies().">DADA2 mapping file</span> <input
              name="dada2_file" accept="text/tab-delimited-values"
              type="file"></li>
				<li><span id="formlabel" tooltip="Select the version of the database that you used to get relative abundances.">Database</span> <input
              name="database" value="midas_v1.0" required="required" type="radio" checked="checked">MIDAS
            1.0 <input name="database" value="midas_v1.2" required="required" 
              type="radio">MIDAS 1.2</li>
				<li><span id="formlabel" tooltip="Select whether you want to calculate prevalence or environmental specificity scores, or provide your own phenotype.">Phenotype</span> <input
              name="phenotype" value="prevalence" required="required" type="radio" checked="checked">prevalence&nbsp;
            <input name="phenotype" value="specificity" required="required" type="radio">environmental
            specificity score <input name="phenotype" value="provided" required="required"
              type="radio">provided </li>
        <li><span id="formlabel" tooltip="If providing your own phenotype, upload it here.">Phenotype file (if
						provided)</span> <input name="phenotype_file" accept="text/tab-delimited-values"
              type="file"></li>
				<li><span id="formlabel" tooltip="The environment for which you are assessing prevalence or specificity. Must match one of the environments in the sample metadata.">Which environment </span>&nbsp;
              <textarea name="which_envir" required="required"
rows="1" maxlength="500">Stool</textarea><br>
          </li>
					<li></ul><span id="bigformlabel">Advanced options</span>
						<ul>
					<li><span id="formlabel" tooltip="Advanced: specify how to calculate the prior probabilities of environments. Most users can leave this set to &quot;uninformative&quot;. You can also weight environments by how diverse they are, or provide your own weighting. For a case-control study, you may wish to provide the actual incidence of the disease in the population, since cases are likely to be oversampled." tooltip-type='advanced'>Environment prior type</span>
            <input name="prior_type" value="uninformative" required="required" type="radio" checked="checked">uninformative
            <input name="prior_type" value="uninformative" required="required" type="radio">alpha-diversity
            <input name="prior_type" value="uninformative" required="required" type="radio">provided</li>
        <li><span id="formlabel" tooltip="Advanced: tab-separated values giving priors. Rows should be sample IDs, and the first data column should give the priors. These will be renormalized to sum to 1.00 if they don't already.">Environment prior file
						(if provided)</span> <input name="prior_file" accept="text/tab-delimited-values"
              type="file"></li>
        <li><span id="formlabel" tooltip="Advanced: only return genes that are present this many times and absent this many times. Filters out associations that depend on only one or two observations.">Minimum number of
						presence/absence values per gene</span>: <input name="minimum"
              value="3" required="required" min="0" max="9999" type="number"></li>
      </ul>
      
<input type="submit" value="Submit" formmethod="post"><br>
    </form>
		<h3>Patrick H. Bradley, Pollard lab</h3>
  </body>
</html>
