---
title: "Phylogenetic linear modeling results"
output: html_notebook
params:
  ncl: 4
  type: "midas"
  out_dir: "output"
  in_dir: "."
  data_dir: "./data"
  abundance_file: "bodysite-averaged.tab"
  metadata_file: "bodysite-metadata.tab"
  phenotype_file: ""
  db_version: "1.0"
  which_phenotype: "specificity"
  which_envir: "Stool"
  prior_type: "uninformative"
  prior_file: ""
  minimum: 3
---


```{r setup}

DDir = params$data_dir
IDir = params$in_dir
ODir = params$out_dir

dir.create(ODir)

source("plm-functions.R")

# Read in the correct gene presence/absence matrix and corresponding tree files
if (params$type == "midas") {
  if (params$db_version == "midas_v1.0") {
    gene.presence <- readRDS(file.path(DDir, "MIDAS-gene-presence-1.0.rds"))
    trees <- readRDS(file.path(DDir, "chronos-overall-trees.rds"))
    phyla <- intersect(names(trees), names(gene.presence))
    taxa <- lapply(trees, function(x) x$tip.label)
    gene.to.fxn <- fread(file.path(DDir, "family.functions"),
                           header = F)
    colnames(gene.to.fxn) <- c("gene", "function")
    taxonomy <- data.frame(fread(file.path(DDir,
                                           "full-taxonomy.csv")),
                           stringsAsFactors = FALSE)[,-1]
  } else {
    stop(paste0("unknown database version ", params$db_version))
  }
}

# Read abundance, metadata files
abd.mtx <- fastread(file.path(IDir, params$abundance_file))
metadata <- read.table(file.path(IDir, params$metadata_file),
                       row.names = 1,
                       header = TRUE)


```

```{r calculate_phenotype, fig.width = 7, fig.height = 20}

# Calculate the phenotype of interest

if (params$which_phenotype == "prevalence") {
  phenotype <- prev.addw(abd.mtx,
                         params$which_envir,
                         metadata)
  phenoLimits <- quantile(unique(phenotype), c(0.2, 0.8))
  phenoColors <- c(low.col = "black", hi.col = "orange2")
} else if (params$which_phenotype == "specificity") {
  if (params$prior_type == "file") {
    prior.file <- read.table(file.path(IDir, params$prior_file))
  } else {
    prior.file <- NULL
  }
  ess <- calc.ess(abd.mtx,
                  params$which_envir,
                  metadata,
                  params$prior_type,
                  prior.file)
  phenotype <- ess$ess
  phenoLimits <- c(phenoP - 1 * sd(phenotype),
                   phenoP + 1 * sd(phenotype))
  phenoP <- logit(c(ess$priors[ess$priors$env == params$which_envir, "prior"]))
  phenoColors <- c(low.col = "slateblue",
                   mid.col = "gray50",
                   high.col = "tomato")
} else if (params$which_phenotype == "provided") {

  } else stop(paste0("don't know how to calculate the phenotype ",
                   params$which_phenotype))



plotted.pheno.trees <- lapply(1:length(trees), function(tn) {
  gg.cont.tree(trees[[tn]],
               phenotype,
               cLimits = phenoLimits,
               colors = phenoColors,
               cName = paste0(names(trees)[tn], ": ", params$which_phenotype))
})
names(plotted.pheno.trees) <- names(trees)
for (phy in names(plotted.pheno.trees)) print(plotted.pheno.trees[[phy]]$tree)

write.table(file=file.path(ODir, "phenotype.tab"), phenotype, sep = '\t')

```


```{r get_associations}
results <- result.wrapper.plm(phyla = phyla,
                              pheno = phenotype,
                              tree = trees,
                              clusters = taxa,
                              proteins = gene.presence, 
                              n.cores = params$ncl)
signif <- make.sigs(results)
signs <- make.signs(results)
pos.sig <- make.pos.sig(sigs = signif, signs = signs)

results.matrix <- Reduce(rbind, lapply(names(results), function(rn) {
  data.frame(phylum = rn, 
             gene = results[[rn]] %>% colnames,
             effect.size = results[[rn]][1,],
             p.value = results[[rn]][2,]
             )
}))

kable(sapply(pos.sig, length), col.names = "significant genes")

pos.sig.descs <- Reduce(rbind, lapply(names(pos.sig), function(n) {
  cbind(phylum = n,
        gene = pos.sig[[n]],
        description = gene.annot(pos.sig[[n]]))
}))

pos.sig.thresh <- mapply(trees, pos.sig, gene.presence,
                         FUN = function(tr, x, y) {
  y2 <- y[x, intersect(tr$tip.label, colnames(y))]
  r1 <- rowSums(y2)
  r2 <- rowSums(1 - y2)
  names(which((r2 >= params$minimum) & (r1 >= params$minimum)))
})

pos.sig.thresh.descs <- Reduce(rbind, lapply(names(pos.sig.thresh),
                                             function(n) {
  cbind(phylum = n,
        gene = pos.sig.thresh[[n]],
        description = gene.annot(pos.sig.thresh[[n]]))
}))

write.csv(file = file.path(ODir, "pos-sig-thresholded.csv"),
          pos.sig.thresh.descs)

write.csv(file = file.path(ODir, "all-results.csv"),
          results.matrix)
```

```{r gene_assoc_heat_maps}
cluster.plots <- mapply(gene.presence, pos.sig.thresh, trees,
                        FUN=function(x, y, z) {
  sig.bin <- x[intersect(rownames(x), y), ]
  names(dimnames(sig.bin)) <- c("gene", "id")
  clust <- hclust(dist(sig.bin, method = "binary"))
  sig.ord <- melt(t(sig.bin[clust$order, ]))
  p <- ggtree(z, ladderize = TRUE)
  facet_plot(p, 'heatmap', sig.ord, geom_tile,
             aes(x = as.numeric(as.factor(gene)),
                 fill = as.numeric(as.factor(value))))
}, SIMPLIFY = FALSE)
for (cp in cluster.plots) print(cp)
```
