---
title: "Phylogenetic linear modeling results"
output: html_notebook
params:
  ncl: 4
  type: "midas"
  out_dir: "output"
  in_dir: "."
  data_dir: "./data"
  abundance_file: "bodysite-averaged.tab"
  metadata_file: "bodysite-metadata.tab"
  phenotype_file: ""
  dada2_file: ""
  db_version: "midas_v1.0"
  which_phenotype: "prevalence"
  which_envir: "Stool"
  prior_type: "uninformative"
  prior_file: ""
  minimum: 3
  burst_dir: "/home/pbradz/bin/"
---

This report was generated automatically by [phylogenize](http://pbradley-dell/~pbradz/phylogenize/) (citation info here), with the following parameters:

  - Type of input data: "`r params$type`"
  - Phenotype calculated: "`r params$which_phenotype`"
  - Environment of interest: "`r params$which_envir`"
  - Prior probabilities for environments: "`r params$prior_type`"


```{r setup, warning=FALSE}

DDir = params$data_dir
IDir = params$in_dir
ODir = params$out_dir
BurstDir = params$burst_dir

dir.create(ODir)

source("plm-functions.R")

# Read in the correct gene presence/absence matrix and corresponding tree files
if (params$type == "midas") {
		SixteenS = FALSE
	if (params$db_version == "midas_v1.0") {
		gene.presence <- readRDS(file.path(DDir, "MIDAS-gene-presence-1.0.rds"))
		trees <- readRDS(file.path(DDir, "MIDAS_1.0-trees.rds"))
		phyla <- intersect(names(trees), names(gene.presence))
		taxonomy <- data.frame(fread(file.path(DDir,
					"full-taxonomy.csv")),
			stringsAsFactors = FALSE)[,-1]
	} else if (params$db_version == "midas_v1.2") {
		gene.presence <- readRDS(file.path(DDir, "MIDAS-gene-presence-1.2.rds"))
		trees <- readRDS(file.path(DDir, "MIDAS_1.2-trees.rds"))
		phyla <- intersect(names(trees), names(gene.presence))
		gene.to.fxn <- fread(file.path(DDir, "family.functions"),
			header = F)
		colnames(gene.to.fxn) <- c("gene", "function")
		taxonomy <- data.frame(fread(file.path(DDir,
					"full-taxonomy.csv")),
			stringsAsFactors = FALSE)[,-1]
	} else {
		stop(paste0("unknown database version ", params$db_version))
	}
} else if (params$type == "16S") {
	SixteenS = TRUE
	gene.presence <- readRDS(file.path(DDir, "MIDAS-gene-presence-1.2.rds"))
	trees <- readRDS(file.path(DDir, "MIDAS_1.2-trees.rds"))
	phyla <- intersect(names(trees), names(gene.presence))
	gene.to.fxn <- fread(file.path(DDir, "family.functions"),
		header = F)
	colnames(gene.to.fxn) <- c("gene", "function")
	taxonomy <- data.frame(fread(file.path(DDir,
				"MIDAS_1.2-taxonomy.csv")),
		stringsAsFactors = FALSE)[,-1]
}

gene.to.fxn <- fread(file.path(DDir, "family.functions"),
	header = F)
colnames(gene.to.fxn) <- c("gene", "function")

fig.hierarchy <- data.frame(fread(file.path(DDir, "subsys.txt"),
		header = F,
	))[,1:4]
colnames(fig.hierarchy) <-  c("level1", "level2", "level3", "function")

g.mappings <- lapply.across.names(colnames(fig.hierarchy), function(x) {
	gene.to.subsys <- merge(data.frame(gene.to.fxn),
		fig.hierarchy[, c(x, "function")],
		by.x = "function.", by.y = "function")[, -1]
})

# Read abundance, metadata files
abd.mtx <- fastread(file.path(IDir, params$abundance_file))
metadata <- read.table(file.path(IDir, params$metadata_file),
                       header = TRUE)
if (SixteenS) {
  # write out input sequences
  write.fasta(as.list(rownames(abd.mtx)),
    paste0("Row", (1:nrow(abd.mtx))),
    file.out = file.path(IDir, "input_seqs.fa"),
    nbchar = 99999,
    as.string = TRUE)
	# map to MIDAS IDs using Burst
  system2(file.path(BurstDir, "burst12"), args = c("-r",
      file.path(DDir, "16s_renamed.frn"),
      "-fr",
      "-q",
      file.path(IDir, "input_seqs.fa"),
      "-i",
      "0.985",
      "-o",
      file.path(IDir, "output_assignments.txt")))
  assignments <- data.frame(fread(file.path(IDir, "output_assignments.txt")))
  row.hits <- as.numeric(gsub("Row", "", assignments[,1]))
  row.targets <- sapply(assignments[,2], function(x) strsplit(x, " ")[[1]][3])
  uniq.hits <- which(count.each(row.hits) < 2)
  rh <- row.hits[uniq.hits]
  rt <- row.targets[uniq.hits]
  subset.abd <- abd.mtx[rh, ]
  urt <- unique(rt)
  summed.uniq <- sapply(urt, function(r) {
    w <- which(rt == r)
    if (length(w) > 1) {
      apply(subset.abd[w, ], 2, sum)
    } else {
      subset.abd[w, ]
    }
  }) %>% t
  rownames(summed.uniq) <- urt
  old.abd.mtx <- abd.mtx
  csu <- colSums(summed.uniq)
  abd.mtx <- apply(summed.uniq[, which(csu > 0)], 2, function(x) x / sum(x))
}

### Figure out how many trees to remain

taxa.observed <- rownames(abd.mtx)
taxa.per.tree <- lapply(trees, function(tr) {
  intersect(tr$tip.label, taxa.observed)
})
saved.phyla <- nw(sapply(taxa.per.tree, length) >= 5)
trees <- trees[saved.phyla]
all.possible.taxa <- Reduce(union, lapply(gene.presence, colnames))
not.observed.taxa <- setdiff(all.possible.taxa, taxa.observed)
taxa.zero <- matrix(rep(0, length(not.observed.taxa) * ncol(abd.mtx)), 
  nr = length(not.observed.taxa),
  byrow = TRUE,
  dimnames = list(not.observed.taxa, colnames(abd.mtx)))
abd.mtx <- rbind(abd.mtx, taxa.zero)
taxa <- lapply(trees, function(x) x$tip.label)

```

First, calculate the phenotype of interest:

```{r calculate_phenotype, fig.width = 7, fig.height = 20, warning = FALSE, message = FALSE}

# Calculate the phenotype of interest

if (params$which_phenotype == "prevalence") {
  phenotype <- prev.addw(abd.mtx,
                         params$which_envir,
                         metadata)
  phenoLimits <- quantile(unique(phenotype), c(0.2, 0.8))
  phenoLimitsPhylum <- lapply(trees, function(tr) {
    phi <- phenotype[intersect(names(phenotype), tr$tip.label)] %>% na.omit
    quantile(unique(phi), c(0.2, 0.8))
  })
  phenoColors <- c(low.col = "black", high.col = "orange2")
} else if (params$which_phenotype == "specificity") {
  if (params$prior_type == "file") {
    prior.file <- read.table(file.path(IDir, params$prior_file))
  } else {
    prior.file <- NULL
  }
  ess <- calc.ess(abd.mtx,
                  params$which_envir,
                  metadata,
                  params$prior_type,
                  prior.file)
  phenotype <- ess$ess
  phenoP <- logit(c(ess$priors[ess$priors$env == params$which_envir, "prior"]))
  phenoLimitsOverall <- c(phenoP - 1 * sd(phenotype),
                          phenoP + 1 * sd(phenotype))
  phenoLimitsPhylum <- lapply(trees, function(tr) {
    phi <- phenotype[intersect(names(phenotype), tr$tip.label)] %>% na.omit %>%
      unique
    c(phenoP - (1 * sd(phi)), phenoP + (1 * sd(phi)))
  })
  phenoColors <- c(low.col = "slateblue",
                   mid.col = "gray50",
                   high.col = "tomato")
} else if (params$which_phenotype == "provided") {

  } else stop(paste0("don't know how to calculate the phenotype ",
                   params$which_phenotype))



plotted.pheno.trees <- lapply(1:length(trees), function(tn) {
  gg.cont.tree(trees[[tn]],
               phenotype,
               cLimits = phenoLimitsPhylum[[tn]],
               colors = phenoColors,
               cName = paste0(names(trees)[tn], ": ", params$which_phenotype))
})
names(plotted.pheno.trees) <- names(trees)
#for (phy in names(plotted.pheno.trees)) print(plotted.pheno.trees[[phy]]$tree)

write.table(file=file.path(ODir, "phenotype.tab"), phenotype, sep = '\t')

```


```{r get_associations, warning = FALSE, message = FALSE}
results <- result.wrapper.plm(phyla = names(taxa),
                              pheno = phenotype,
                              tree = trees,
                              clusters = taxa,
                              proteins = gene.presence, 
                              n.cores = params$ncl)
signif <- make.sigs(results)
signs <- make.signs(results)
pos.sig <- make.pos.sig(sigs = signif, signs = signs)

results.matrix <- Reduce(rbind, lapply(names(results), function(rn) {
  data.frame(phylum = rn, 
             gene = results[[rn]] %>% colnames,
             effect.size = results[[rn]][1,],
             p.value = results[[rn]][2,]
             )
}))

kable(sapply(pos.sig, length), col.names = "significant genes")

pos.sig.descs <- Reduce(rbind, lapply(names(pos.sig), function(n) {
  cbind(phylum = n,
        gene = pos.sig[[n]],
        description = gene.annot(pos.sig[[n]]))
}))

pos.sig.thresh <- mapply(trees, pos.sig, gene.presence[names(trees)],
                         FUN = function(tr, x, y) {
  y2 <- y[x, intersect(tr$tip.label, colnames(y))]
  r1 <- rowSums(y2)
  r2 <- rowSums(1 - y2)
  names(which((r2 >= params$minimum) & (r1 >= params$minimum)))
})

pos.sig.thresh.descs <- Reduce(rbind, lapply(names(pos.sig.thresh),
                                             function(n) {
  cbind(phylum = n,
        gene = pos.sig.thresh[[n]],
        description = gene.annot(pos.sig.thresh[[n]]))
}))

write.csv(file = file.path(ODir, "pos-sig-thresholded.csv"),
          pos.sig.thresh.descs)

write.csv(file = file.path(ODir, "all-results.csv"),
          results.matrix)
```

```{r gene_assoc_heat_maps, warning = FALSE, message = FALSE}
cluster.plots <- mapply(gene.presence[names(trees)], pos.sig.thresh, trees,
                        plotted.pheno.trees, names(trees),
                        FUN=function(x, y, z, w, nt) {
  sig.bin <- x[intersect(rownames(x), y), ]
  names(dimnames(sig.bin)) <- c("gene", "id")
  clust <- hclust(dist(sig.bin, method = "binary"))
  sig.ord <- melt(t(sig.bin[clust$order, ]))
  col.df <- data.frame(id = names(w$disp), disp = w$disp)
  p <- ggtree(z, ladderize = TRUE) %<+% col.df +
    geom_tippoint(aes(color = disp, shape = '.'))
  if (params$which_phenotype == "prevalence") {
    p <- p + scale_color_gradient(low = w$cols["low.col"],
                                  high = w$cols["high.col"])
  } else if (params$which_phenotype == "specificity") {
     p <- p + scale_color_gradient2(low = w$cols["low.col"],
                                    mid = w$cols["mid.col"],
                                  high = w$cols["high.col"],
                                  midpoint = mean(w$lims))   
  }
  tmp <- facet_plot(p,
                    paste0('heatmap: ', nt),
                    sig.ord,
                    geom_tile,
                    aes(x = as.numeric(as.factor(gene)),
                        fill = as.numeric(as.factor(value)))) +
    scale_fill_gradient(low = "#000000", high = "slateblue2")
  print(tmp)
  return(tmp)
}, SIMPLIFY = FALSE)
```

Enrichment analysis


```{r enrichment_analysis, warning = FALSE, message = FALSE}
enrichments <- perform.enrichments(overall.sigs, 
	overall.signs,
	overall.results,
	mapping = g.mappings,
	verbose = FALSE)
enr.table <- make.plm.enr.tables(enrichments)
write.csv(file=file.path(ODir, "enr-table.csv"), enr.table)
kable(enr.table)
```
