---
title: "Phylogenetic linear modeling results"
output: 
  html_document:
    toc: true
    toc_float: true
    keep_md: true
    self_contained: true
params:
  ncl: 1
  type: "midas"
  source_dir: "."
  out_dir: "."
  in_dir: "."
  data_dir: "./data"
  abundance_file: "abundance.tab"
  metadata_file: "metadata.tab"
  biom_file: ""
  input_format: "tabular"
  env_column: "env"
  taxon_level: "phylum"
  dset_column: "dataset"
  phenotype_file: ""
  db: "gtdb"
  which_phenotype: "prevalence"
  which_envir: "Stool"
  prior_type: "uninformative"
  prior_file: ""
  minimum: 3
  treemin: 5
  assume_below_LOD: TRUE
  skip_graphs: FALSE
  vsearch_dir: "./bin"
  core_method: "phylogenize"
  use_rmd_params: TRUE
  error_to_file: TRUE
---

```{r setup, warning=FALSE, message=FALSE, results="hide", echo=FALSE, cache=FALSE}

library(phylogenize)
library(Matrix)

cache.d <- file.path(pz.options('out_dir'), 'cache')
dir.create(pz.options('out_dir'))
dir.create(cache.d)
knitr::opts_chunk$set(cache.path=cache.d)

if (pz.options('use_rmd_params')) {
    do.call(pz.options, args=params)
}

if (is.null(pz.options('relative_out_dir'))) {
  pz.options(relative_out_dir=pz.options('out_dir'))
}

# Note that params give defaults for the web tool, while PZ_OPTIONS gives
# defaults for the package

# set graphing hook
if (!pz.options('skip_graphs')) {
    library(svglite)
    knitr::knit_hooks$set(custom.plot = knitr::hook_plot_custom)
    # Thanks to Yihui Xie:
    # https://github.com/yihui/knitr/issues/754#issuecomment-40540980
    local({
        hook_plot = knitr::knit_hooks$get('plot')
        knitr::knit_hooks$set(plot = function(x, options) {
            x = paste(x, collapse = '.')
            if (!grepl('\\.svg', x)) return(hook_plot(x, options))
            if (file.exists(x)) {
              paste(readLines(x)[-1], collapse = '\n')
            } else {
              paste("<!-- file not found: ",
                    x,
                    " -->",
                    sep = '',
                    collapse = '')
            }
        })
    })
}

# Only load lobstr to track memory usage if debugging
dir.create(pz.options('out_dir'))
do_POMS <- FALSE
if (tolower(pz.options('core_method'))=="phylogenize") {
  p.method <- phylolm.fx.pv
} else if (tolower(pz.options('core_method'))=="poms") {
  p.method <- NULL
  do_POMS <- TRUE
} else if (tolower(pz.options('core_method'))=="linear") {
  p.method <- lm.fx.pv
} else {
  pz.error("core_method must be phylogenize, poms, or linear")
}

# Read in user-supplied
abd.meta <- read.abd.metadata()
# Read in trees, gene presence/absence, taxonomy
pz.db <- import.pz.db()
# Figure out how many trees to retain
pz.db <- adjust.db(pz.db, abd.meta)
```

# Settings

This report was generated automatically by [phylogenize](https://phylogenize.org/) (Bradley and Pollard, 2019: https://phylogenize.org), with the following parameters:

- Type of input data: "`r pz.options('db')`"
- Phenotype calculated: "`r pz.options('which_phenotype')`"
- Environment of interest: "`r pz.options('which_envir')`"
- Prior probabilities for environments: "`r pz.options('prior_type')`"

# Phenotypic Calculation

Below, trees for all taxa with at least `r pz.options('treemin')`
representatives in the input data are presented. Trees are colored according to
the phenotype chosen (here, `r pz.options('which_phenotype')` in the environment
 `r pz.options('which_envir') `).

For prevalence, the tree is colored black (low prevalence) to orange (high
prevalence). For specificity, the tree is colored red for taxa that are
specific for an environment, blue for taxa that are specific for other
environments (i.e., found less often in the environment chosen), and gray for
taxa that are neither.

## Distribution of Phenotype {.tabset}
```{r calculate_phenotype, fig.width = 7, fig.height = 20, warning = FALSE, message = FALSE, results="hide", echo=FALSE, cache=FALSE}

# Read in user-supplied data, trees, gene info, taxonomy; calculate phenotypes;
# and automatically unpack results to current environment
# Only save the abd.meta data if we're running POMS, which needs it...
list2env(data_to_phenotypes(save_data = !do_POMS), environment())
# Also unpack "phenotype, phenoP, ..." into current environment
list2env(phenotype_results, environment())

mapped.observed <- names(which(Matrix::rowSums(abd.meta$mtx) > 0))

phenotype_info <- calc.phenotype.interest(abd.meta, pz.db, mapped.observed)
pz.db <- phenotype_info[[1]]
phenotype <- phenotype_info[[2]]
phenoP <- phenotype_info[[3]]

phenotype <- clean.pheno(phenotype, pz.db)
pheno.distros <- plot.pheno.distributions(phenotype, pz.db)
pheno.scale <- get.pheno.plotting.scales(phenotype, pz.db$trees, phenoP)
plotted.pheno.trees <- plot.phenotype.trees(phenotype, pz.db$trees, pz.db$taxonomy, pheno.scale)

# Template for plots
template <- c(
    "### Page {{nm}}\n",
    "```{r, echo = FALSE, warning = FALSE, message = FALSE}\n",
    "pheno.distros[[{{nm}}]]\n",
    "```\n",
    "\n"
)

plots <- lapply(1:length(pheno.distros), function(nm) {
    knitr::knit_expand(text = template, nm = nm)
})

write.table(file=file.path(pz.options('out_dir'), "phenotype.tab"),
            phenotype,
            sep = '\t')

# be a little parsimonious with memory here
gc()
```
`r knitr::knit_child(text = unlist(plots), quiet = TRUE)`
## Phenotype Mapped Across Trees {.tabset}
### All Taxa
```{r plot_pheno_trees, echo=FALSE, warning = FALSE, results='asis'}

plots <- list()
if (!pz.options('skip_graphs')) {
	plots <- plot.labeled.phenotype.trees(plotted.pheno.trees,
				     100*logistic(phenotype),
				     label=pz.options('which_phenotype'))
}

clinUtils::knitPrintListPlots(
	plotsList = plots, 
	type = "plotly", 
	titles = NULL, 
	titleLevel = NULL
)

### Only mapped/observed taxa {.tabset}
```{r subtrees_pre, cache=FALSE, echo=FALSE}

pheno.labels <- names(phenotype)
subtrees <- lapply(pz.db$trees, function(tr) {
  keep.tips(tr, intersect(mapped.observed, pheno.labels))
})

subtrees <- Filter(function(x) {
  if (is.null(x)) return(FALSE)
  if (length(x$tip.label) > 2) return(TRUE)
  return(FALSE)
}, subtrees)

```
```{r subtrees, fig.width = 7, fig.height = 20, warning = FALSE, message = FALSE, results="hide", echo=FALSE, fig.keep = 'none', fig.ext='svg', custom.plot=TRUE, fig.num=length(subtrees), cache.lazy = FALSE, cache=FALSE, echo=FALSE}

plots <- list()
if (!pz.options('skip_graphs')) {
  plots <- plot.phenotype.trees(phenotype,
                                                 subtrees,
						 pz.db$taxonomy,
                                                 pheno.scale)
  if (!is.null(plotted.pheno.subtrees)) {
    if (length(plotted.pheno.subtrees) > 0) {
      plot.labeled.phenotype.trees(plotted.pheno.subtrees,
                                   100*logistic(phenotype))
      rmd.message <- ""
    } else {
      rmd.message <- "(no taxa had more than one tip)"
    }
  } else {
    rmd.message <- "(no taxa had more than one tip)"
  }
}

clinUtils::knitPrintListPlots(
        plotsList = plots,
        type = "plotly",
        titles = NULL,
        titleLevel = NULL
)
