---
title: "Phylogenetic linear modeling results"
output: 
  html_document:
    toc: true
    toc_float: true
    keep_md: true
    self_contained: true
params:
  ncl: 1
  type: "midas"
  source_dir: "."
  out_dir: "."
  in_dir: "."
  data_dir: "./data"
  abundance_file: "abundance.tab"
  metadata_file: "metadata.tab"
  biom_file: ""
  input_format: "tabular"
  env_column: "env"
  taxon_level: "phylum"
  dset_column: "dataset"
  phenotype_file: ""
  db: "gtdb"
  which_phenotype: "prevalence"
  which_envir: "Stool"
  prior_type: "uninformative"
  prior_file: ""
  minimum: 3
  treemin: 5
  assume_below_LOD: TRUE
  skip_graphs: FALSE
  vsearch_dir: "./bin"
  linearize: FALSE
  use_rmd_params: TRUE
  error_to_file: TRUE
---

```{r setup, warning=FALSE, message=FALSE, results="hide", echo=FALSE, cache=FALSE}

library(phylogenize)
library(Matrix)

cache.d <- file.path(pz.options('out_dir'), 'cache')
dir.create(pz.options('out_dir'))
dir.create(cache.d)
knitr::opts_chunk$set(cache.path=cache.d)

if (pz.options('use_rmd_params')) {
    do.call(pz.options, args=params)
}

if (is.null(pz.options('relative_out_dir'))) {
  pz.options(relative_out_dir=pz.options('out_dir'))
}

# Note that params give defaults for the web tool, while PZ_OPTIONS gives
# defaults for the package

# set graphing hook
if (!pz.options('skip_graphs')) {
    library(svglite)
    knitr::knit_hooks$set(custom.plot = knitr::hook_plot_custom)
    # Thanks to Yihui Xie:
    # https://github.com/yihui/knitr/issues/754#issuecomment-40540980
    local({
        hook_plot = knitr::knit_hooks$get('plot')
        knitr::knit_hooks$set(plot = function(x, options) {
            x = paste(x, collapse = '.')
            if (!grepl('\\.svg', x)) return(hook_plot(x, options))
            if (file.exists(x)) {
              paste(readLines(x)[-1], collapse = '\n')
            } else {
              paste("<!-- file not found: ",
                    x,
                    " -->",
                    sep = '',
                    collapse = '')
            }
        })
    })
}

# Only load lobstr to track memory usage if debugging
dir.create(pz.options('out_dir'))

# Read in user-supplied
abd.meta <- read.abd.metadata()
# Read in trees, gene presence/absence, taxonomy
pz.db <- import.pz.db()
# Figure out how many trees to retain
pz.db <- adjust.db(pz.db, abd.meta)

pz.db$trees <- Filter(function(tr) length(tr$tip.label) > 1, pz.db$tree)

if (pz.options('assume_below_LOD')) {
  abd.meta <- add.below.LOD(pz.db, abd.meta)
  sanity.check.abundance(abd.meta$mtx)
}
# Make back into a binary sparse matrix to save memory
pz.message("Starting evaluation...")
```

# Settings

This report was generated automatically by [phylogenize](https://phylogenize.org/) (Bradley and Pollard, 2019: https://phylogenize.org), with the following parameters:

- Type of input data: "`r pz.options('db')`"
- Phenotype calculated: "`r pz.options('which_phenotype')`"
- Environment of interest: "`r pz.options('which_envir')`"
- Prior probabilities for environments: "`r pz.options('prior_type')`"

# Phenotypic Calculation

Below, trees for all taxa with at least `r pz.options('treemin')`
representatives in the input data are presented. Trees are colored according to
the phenotype chosen (here, `r pz.options('which_phenotype')` in the environment
 `r pz.options('which_envir') `).

For prevalence, the tree is colored black (low prevalence) to orange (high
prevalence). For specificity, the tree is colored red for taxa that are
specific for an environment, blue for taxa that are specific for other
environments (i.e., found less often in the environment chosen), and gray for
taxa that are neither.

## Distribution of Phenotype {.tabset}
```{r calculate_phenotype, fig.width = 7, fig.height = 20, warning = FALSE, message = FALSE, results="hide", echo=FALSE, cache=FALSE}

# Calculate the phenotype of interest

mapped.observed <- names(which(Matrix::rowSums(abd.meta$mtx) > 0))

phenotype_info <- calc.phenotype.interest(abd.meta, pz.db)
pz.db <- phenotype_info[[1]]
phenotype <- phenotype_info[[2]]
phenoP <- phenotype_info[[3]]

phenotype <- clean.pheno(phenotype, pz.db)
pheno.distros <- plot.pheno.distributions(phenotype, pz.db)
pheno.scale <- get.pheno.plotting.scales(phenotype, pz.db$trees, phenoP)
plotted.pheno.trees <- plot.phenotype.trees(phenotype, pz.db$trees, pz.db$taxonomy, pheno.scale)

# Template for plots
template <- c(
    "### Page {{nm}}\n",
    "```{r, echo = FALSE, warning = FALSE, message = FALSE}\n",
    "pheno.distros[[{{nm}}]]\n",
    "```\n",
    "\n"
)

plots <- lapply(1:length(pheno.distros), function(nm) {
    knitr::knit_expand(text = template, nm = nm)
})

write.table(file=file.path(pz.options('out_dir'), "phenotype.tab"),
            phenotype,
            sep = '\t')

```
`r knitr::knit_child(text = unlist(plots), quiet = TRUE)`
## Phenotype Mapped Across Trees {.tabset}
### All Taxa
```{r plot_pheno_trees, echo=FALSE, warning = FALSE, results='asis'}

plots <- list()
if (!pz.options('skip_graphs')) {
	plots <- plot.labeled.phenotype.trees(plotted.pheno.trees,
				     100*logistic(phenotype),
				     label=pz.options('which_phenotype'))
}

# with titles
clinUtils::knitPrintListPlots(
	plotsList = plots, 
	type = "plotly", 
	titles = names(plots), 
	titleLevel = 3
)
