---
title: "Phylogenetic linear modeling results"
output: 
  html_document:
    toc: true
    toc_float: true
    keep_md: true
    self_contained: true
params:
  ncl: 1
  type: "midas"
  source_dir: "."
  out_dir: "."
  in_dir: "."
  data_dir: "./data"
  abundance_file: "abundance.tab"
  metadata_file: "metadata.tab"
  biom_file: ""
  input_format: "tabular"
  env_column: "env"
  dset_column: "dataset"
  phenotype_file: ""
  db: "gtdb"
  which_phenotype: "prevalence"
  which_envir: "Stool"
  prior_type: "uninformative"
  prior_file: ""
  minimum: 3
  treemin: 5
  assume_below_LOD: TRUE
  skip_graphs: FALSE
  vsearch_dir: "./bin"
  linearize: FALSE
  use_rmd_params: TRUE
  error_to_file: TRUE
---

```{r setup, warning=FALSE, message=FALSE, results="hide", echo=FALSE, cache=FALSE}

library(phylogenize)
library(Matrix)

cache.d <- file.path(pz.options('out_dir'), 'cache')
dir.create(pz.options('out_dir'))
dir.create(cache.d)
knitr::opts_chunk$set(cache.path=cache.d)

if (pz.options('use_rmd_params')) {
    do.call(pz.options, args=params)
}

if (is.null(pz.options('relative_out_dir'))) {
  pz.options(relative_out_dir=pz.options('out_dir'))
}

# Note that params give defaults for the web tool, while PZ_OPTIONS gives
# defaults for the package

# set graphing hook
if (!pz.options('skip_graphs')) {
    library(svglite)
    knitr::knit_hooks$set(custom.plot = knitr::hook_plot_custom)
    # Thanks to Yihui Xie:
    # https://github.com/yihui/knitr/issues/754#issuecomment-40540980
    local({
        hook_plot = knitr::knit_hooks$get('plot')
        knitr::knit_hooks$set(plot = function(x, options) {
            x = paste(x, collapse = '.')
            if (!grepl('\\.svg', x)) return(hook_plot(x, options))
            if (file.exists(x)) {
              paste(readLines(x)[-1], collapse = '\n')
            } else {
              paste("<!-- file not found: ",
                    x,
                    " -->",
                    sep = '',
                    collapse = '')
            }
        })
    })
}

# Only load lobstr to track memory usage if debugging
dir.create(pz.options('out_dir'))

# Read in user-supplied
abd.meta <- read.abd.metadata()
# Read in trees, gene presence/absence, taxonomy
pz.db <- import.pz.db()
# Figure out how many trees to retain
pz.db <- adjust.db(pz.db, abd.meta)
if (pz.options('assume_below_LOD')) {
  abd.meta <- add.below.LOD(pz.db, abd.meta)
  sanity.check.abundance(abd.meta$mtx)
}
# Make back into a binary sparse matrix to save memory
pz.message("Starting evaluation...")

```

# Settings

This report was generated automatically by [phylogenize](https://phylogenize.org/) (Bradley and Pollard, 2019: https://phylogenize.org), with the following parameters:

- Type of input data: "`r pz.options('db')`"
- Phenotype calculated: "`r pz.options('which_phenotype')`"
- Environment of interest: "`r pz.options('which_envir')`"
- Prior probabilities for environments: "`r pz.options('prior_type')`"

# Phenotype calculation

Below, trees for all taxa with at least `r pz.options('treemin')`
representatives in the input data are presented. Trees are colored according to
the phenotype chosen (here, `r pz.options('which_phenotype')` in the environment
 `r pz.options('which_envir') `).

For prevalence, the tree is colored black (low prevalence) to orange (high
prevalence). For specificity, the tree is colored red for taxa that are
specific for an environment, blue for taxa that are specific for other
environments (i.e., found less often in the environment chosen), and gray for
taxa that are neither.



```{r calculate_phenotype, fig.width = 7, fig.height = 20, warning = FALSE, message = FALSE, results="hide", echo=FALSE, cache=FALSE}

# Calculate the phenotype of interest

mapped.observed <- names(which(Matrix::rowSums(abd.meta$mtx) > 0))

if (pz.options('which_phenotype') == "prevalence") {
  phenotype <- prev.addw(abd.meta)
  phenoP <- NULL
} else if (pz.options('which_phenotype') == "specificity") {
  if (pz.options('prior_type') == "file") {
    prior.data <- read.table(file.path(pz.options('input_dir'),
                                       pz.options('prior_file')))
  } else {
    prior.data <- NULL
  }
  ess <- calc.ess(abd.meta,
                  prior.data)
  phenotype <- ess$ess
  phenoP <- ess$phenoP
  # Specifically for specificity, only retain observed taxa
  pz.db$trees <- retain.observed.taxa(pz.db$trees,
                                      phenotype,
                                      phenoP,
                                      mapped.observed)
  pz.db$species <- lapply(pz.db$trees, function(x) x$tip.label)
  pz.db$ntaxa <- length(pz.db$trees)
} else if (pz.options("which_phenotype") == "provided") {
  p_tbl <- read_tsv(pz.options("phenotype_file"))
  if (ncol(p_tbl) == 2) { # assume we only have species IDs and values
    phenotype <- deframe(p_tbl)
  } else { # perform shrinkage on the provided values w/ their stderrs
    p_est <- as.numeric(p_tbl[["estimate"]])
    p_se <- as.numeric(p_tbl[["stderr"]])
    names(p_est) <- p_tbl[[1]]
    names(p_se) <- p_tbl[[1]]
    ashr_res <- ashr_wrapper(p_est, p_se)
    phenotype <- ashr_res$result %>%
      as_tibble(rownames="species") %>%
      dplyr::select(species, PosteriorMean) %>%
      deframe
  }
} else if (pz.options("which_phenotype") == "abundance") {
  phenotype <- ashr.diff.abund(abd.meta)
  phenoP <- 0
} else {
  pz.error(paste0("don't know how to calculate the phenotype ",
                  pz.options('which_phenotype')))
}

phenotype <- clean.pheno(phenotype, pz.db)
pheno.scale <- get.pheno.plotting.scales(phenotype, pz.db$trees, phenoP)
plotted.pheno.trees <- plot.phenotype.trees(phenotype,
                                            pz.db$trees,
                                            pheno.scale)
pheno.distros <- plot.pheno.distributions(phenotype,
                                          pz.db)

print(pheno.distros)
write.table(file=file.path(pz.options('out_dir'), "phenotype.tab"),
            phenotype,
            sep = '\t')

# be a little parsimonious with memory here
rm(abd.meta); gc()
```

# Phenotype mapped across trees {.tabset}

## All taxa

```{r plot_pheno_trees, fig.width = 7, fig.height = 20, warning = FALSE, message = FALSE, results="hide", echo=FALSE, fig.keep = 'none', fig.ext='svg', custom.plot=TRUE, fig.num=length(plotted.pheno.trees), cache.lazy = FALSE, cache=FALSE}

if (!pz.options('skip_graphs')) {
    plot.labeled.phenotype.trees(plotted.pheno.trees,
                                 100*logistic(phenotype),
                                 label=pz.options('which_phenotype'))
}

```

## Only mapped/observed taxa {.tabset}

Displays phenotype on trees, with only the mapped/observed taxa plotted.

```{r subtrees_pre, cache=FALSE, echo=FALSE}

pheno.labels <- names(phenotype)
subtrees <- lapply(pz.db$trees, function(tr) {
  keep.tips(tr, intersect(mapped.observed, pheno.labels))
})
subtrees <- Filter(function(x) {
  if (is.null(x)) return(FALSE)
  if (length(x$tip.label) > 2) return(TRUE)
  return(FALSE)
}, subtrees)

```

```{r subtrees, fig.width = 7, fig.height = 20, warning = FALSE, message = FALSE, results="hide", echo=FALSE, fig.keep = 'none', fig.ext='svg', custom.plot=TRUE, fig.num=length(subtrees), cache.lazy = FALSE, cache=FALSE, echo=FALSE}

if (!pz.options('skip_graphs')) {
  plotted.pheno.subtrees <- plot.phenotype.trees(phenotype,
                                                 subtrees,
                                                 pheno.scale)
  if (!is.null(plotted.pheno.subtrees)) {
    if (length(plotted.pheno.subtrees) > 0) {
      plot.labeled.phenotype.trees(plotted.pheno.subtrees,
                                   100*logistic(phenotype))
      rmd.message <- ""
    } else {
      rmd.message <- "(no taxa had more than one tip)"
    }
  } else {
    rmd.message <- "(no taxa had more than one tip)"
  }
}

```

`r rmd.message`

# Associations between genes and taxa

This table gives an overview of how many genes were significantly associated at
an FDR of 5%.

```{r get_associations, warning = FALSE, message = FALSE, echo=FALSE, cache=FALSE}

if (pz.options('linearize')) { # mostly for testing
  p.method <- lm.fx.pv
} else {
  p.method <- phylolm.fx.pv
}


taxaN <- names(which(pheno_nonzero_var(phenotype, pz.db$species)))
if (pz.options('ncl') > 1) {
    results <- result.wrapper.plm(taxa=taxaN,
                                  pheno=phenotype,
                                  tree=pz.db$trees[taxaN],
                                  clusters=pz.db$species[taxaN],
                                  proteins=pz.db$gene.presence[taxaN],
                                  method=p.method)
} else {
    results <- mapply(nonparallel.results.generator,
                      pz.db$gene.presence[taxaN],
                      pz.db$trees[taxaN],
                      pz.db$species[taxaN],
                      as.list(taxaN),
                      MoreArgs=list(pheno=phenotype,
                                    method=p.method,
                                    use.for.loop=FALSE),
                      SIMPLIFY=FALSE)
}
signif <- make.sigs(results)
signs <- make.signs(results)

pos.sig <- nonequiv.pos.sig(results, min_fx=pz.options('min_fx'))
results.matrix <- make.results.matrix(results)

knitr::kable(sapply(pos.sig, length), col.names = "significant genes")

phy.with.sigs <- names(which(sapply(pos.sig, length) > 0))
pos.sig.descs <- add.sig.descs(phy.with.sigs, pos.sig, pz.db$gene.to.fxn)

pos.sig.thresh <- threshold.pos.sigs(pz.db, phy.with.sigs, pos.sig)
pos.sig.thresh.descs <- add.sig.descs(phy.with.sigs, pos.sig.thresh, pz.db$gene.to.fxn)

# recalculate, since some of these may go away
phy.with.sigs <- names(which(sapply(pos.sig.thresh, length) > 0))

write.csv(file = file.path(pz.options('out_dir'), "pos-sig-thresholded.csv"),
          pos.sig.thresh.descs)
rm(pos.sig.thresh.descs)

write.csv(file = file.path(pz.options('out_dir'), "all-results.csv"),
          results.matrix)
```

# Gene association heat maps

The left shows the phenotype painted on tips of the tree as above. The right
adds a heatmap where blue points represent presence of a gene (aligned with the
tips of the tree) and white points represent absence. Only
significantly-associated genes are displayed. Genes are clustered from left to
right by Jaccard dissimilarity, so genes with similar patterns of inheritance
will form vertical stripes. Horizontal stripes may be visible for taxa that
have particularly extreme phenotypes.

```{r gene_assoc_heat_maps, warning = FALSE, message = FALSE, echo=FALSE, results="hide", dev = "png", dpi = 300, cache=FALSE}



cluster.plots <- mapply(do.clust.plot, 
                        pz.db$gene.presence[phy.with.sigs],
                        pos.sig.thresh[phy.with.sigs],
                        pz.db$trees[phy.with.sigs],
                        plotted.pheno.trees[phy.with.sigs],
                        phy.with.sigs,
                        SIMPLIFY = FALSE)

rm(cluster.plots)
rm(plotted.pheno.trees)

```

# Enrichment analysis

This process takes the set of genes that are significant and positively
associated with the phenotype, then asks whether there are more than expected
by chance in a given KEGG module or pathway. This requires clusterProfiler to
be installed. </p>

All pathways/modules returned are significant at an FDR of 25%. "Strong", "Med",
and "Weak" refer to per-gene significance cutoffs of 5%, 10%, and 25%,
respectively.

```{r enrichment_analysis, warning = FALSE, message = FALSE, echo = FALSE,results="hide", cache=FALSE}
do_enrichment <- ("clusterProfiler" %in% rownames(installed.packages()))
if (do_enrichment) {
  pretty.enr.tbl <- NULL
  enrichment.tbl <- multi.kegg.enrich(signif,
                                      signs,
                                      pz.db$gene.to.fxn)
  if (!is.null(enrichment.tbl)) {
    enrichment.tbl <- filter(enrichment.tbl,
                             qvalue <= 0.25, enr.estimate > 1)
    pretty.enr.tbl <- select(enrichment.tbl,
                             taxon,
                             cutoff,
                             ID,
                             Description,
                             qvalue,
                             enr.estimate) %>%
       rename(Gene_significance=cutoff,
             Taxon=taxon,
             Description=Description,
             q_value=qvalue,
             Enrichment_odds_ratio=enr.estimate) %>%
      arrange(factor(Gene_significance, levels=names(signif[[1]])),
              Taxon,
              q_value)
    write.csv(file=file.path(pz.options('out_dir'), "enr-table.csv"), pretty.enr.tbl)
  }
}
```
Enrichment tables:

```{r enrichment_display, echo=FALSE, results='asis'}

check_combination <- function(taxon, gene) {
  tryCatch({
    value <- results.matrix[[taxon]][1, gene]
    if (length(value) == 0) {
      return(FALSE)
    }
    return(TRUE)
  }, error = function(e) {
    return(FALSE)
  })
}

options(knitr.table.format = "html")
if (do_enrichment) {
  if (!is.null(pretty.enr.tbl)) {
    cat(output.enr.table(pretty.enr.tbl))
    accession_to_fxn <- pz.db$gene.to.fxn %>%
      select(accession, `function`) %>%
      distinct()
    if (nrow(enrichment.tbl) > 0) {
      enr.overlap <- select(enrichment.tbl,
                            taxon,
                            cutoff,
                            ID,
                            Description,
                            geneID) %>%
        separate_longer_delim(geneID, delim="/") %>%
        left_join(.,
                  accession_to_fxn,
                  by=c("geneID"="accession")) %>%
        rename(gene=geneID, description=`function`) %>%
	mutate(effectsize=map2_dbl(taxon,
                                   gene,
                                   ~{
					   tryCatch({
						   value <- results.matrix[[.x]][1, .y]
					   	   if (length(value) == 0) {
							   return(NA_real_)
						   }
						   return(value)
					   }, error = function(e) {
						   return(NA_real_)
					   })
				     }))
    write.csv(file=file.path(pz.options('out_dir'),
                             "enr-overlaps.csv"),
              enr.overlap)
    write.csv(file=file.path(pz.options('out_dir'),
                             "enr-overlaps-sorted.csv"),
              arrange(enr.overlap,
                      taxon,
                      factor(cutoff, levels=names(signif[[1]])),
                      desc(effectsize)))
    rm(enr.overlap)
    rm(enrichments)
    rm(results.matrix)
  }
}
} else {
  pz.message("Skipping enrichment as clusterProfiler could not be found")
}
```

```{r cleanup, warning=FALSE, message=FALSE, echo=FALSE}
# Clean up memory before rendering
rm(results)
rm(plotted.pheno.subtrees)
rm(plotted.pheno.trees)
rm(pz.db)
```

# Full output files

Links to other files generated by the report:

```{r, results='asis', echo=FALSE}

if (is.null(pz.options('relative_out_dir'))) {
  r.outdir <- pz.options('out_dir')
} else {
  r.outdir <- pz.options('relative_out_dir')
}
for (f in c("phenotype.tab",
            "all-results.csv",
            "pos-sig-thresholded.csv",
            "enr-table.csv",
            "enr-overlaps.csv",
            "enr-overlaps-sorted.csv")) {
  fp <- file.path(r.outdir, f)
  # Note, check the actual output directory when we're trying to figure out what
  # got made
  if (file.exists(file.path(pz.options('out_dir'), f))) {
    cat(paste0(' * [**', f, '**](', fp, ')\n'))
  }
}
```
